### PR TITLE
Remove some allocations in apply and LOBPCG

### PR DESCRIPTION
**Pull request by mfherbst**
Attempt to speed up the apply of the Hamiltonian for bigish calculations. Does two things:
- Removes the allocation of two `fft_size` grid arrays on each apply by preallocating them on Hamiltonian construction
- Cleanup of LOBPCG, removal of some `hcat`s only relevant for the general eigensolver case, usage of a a few `.=` instead of `=` to avoid allocation.

### CODE REVIEW

**Comment by mfherbst** [comment]
I should probably mention that this is not really doing an aweful lot. The biggest issue at the moment is that the subspace routines (mainly the orthogonalisation via Cholesky) is not properly parallelised. Even on big systems hardly more than a few CPU cores are actually employed.

File: src/eigen/lobpcg_hyper_impl.jl [hunk 1]
```diff
@@ -227,64 +227,71 @@ function LOBPCG(A, X, B=I, precon=((Y, X, R)->R), tol=1e-10, maxiter=100; ortho_
 
             # Form Rayleigh-Ritz subspace
             if niter > 2
-                Y = hcat(X,R,P)
-                AY = hcat(AX,AR,AP)
-                BY = hcat(BX,BR,BP)
+                Y = hcat(X, R, P)
+                AY = hcat(AX, AR, AP)
+                if B != I
```
Comment (antoine-levitt) [inline src/eigen/lobpcg_hyper_impl.jl:232]: I would do a ternary op here and below

File: src/eigen/lobpcg_hyper_impl.jl [hunk 2]
```diff
@@ -227,64 +227,71 @@ function LOBPCG(A, X, B=I, precon=((Y, X, R)->R), tol=1e-10, maxiter=100; ortho_
...
-                new_BX = BY*cX
+                new_BX = BY * cX
             else
                 new_BX = new_X
             end
         end
 
         ### Compute new residuals
-        R = new_AX - new_BX*Diagonal(λs)
+        new_R = new_AX - new_BX * Diagonal(λs)
```
Comment (antoine-levitt) [inline src/eigen/lobpcg_hyper_impl.jl:263]: why the rename?

File: src/eigen/lobpcg_hyper_impl.jl [hunk 3]
```diff
@@ -355,12 +362,16 @@ function LOBPCG(A, X, B=I, precon=((Y, X, R)->R), tol=1e-10, maxiter=100; ortho_
         # Orthogonalize R wrt all X, newly active P
         if niter > 1
             Z = hcat(full_X, P)
-            BZ = hcat(full_BX, BP)
+            if B != I
```
Comment (antoine-levitt) [inline src/eigen/lobpcg_hyper_impl.jl:365]: ternary here also

File: src/eigen/lobpcg_hyper_impl.jl [hunk 4]
```diff
@@ -370,11 +381,8 @@ function LOBPCG(A, X, B=I, precon=((Y, X, R)->R), tol=1e-10, maxiter=100; ortho_
             mul!(BR, B, R)
             O = Hermitian(R'*BR)
             U = cholesky(O).U
-            R = R/U
-            BR = BR/U
-            # condU = 1/LAPACK.trcon!('I', 'U', 'N', Array(U))
-        else
-            BR = R
+            R .= R / U
```
Comment (antoine-levitt) [inline src/eigen/lobpcg_hyper_impl.jl:384]: `rdiv!`

File: src/eigen/lobpcg_hyper_impl.jl (hunk 2 above)
Comment (mfherbst) [reply src/eigen/lobpcg_hyper_impl.jl:263]: no not mix it up with the old `R` (was needed for the lazy concatenation that I tried) also follows the analogy with P and X.

File: src/terms/Hamiltonian.jl [hunk 5]
```diff
@@ -11,7 +11,10 @@ struct HamiltonianBlock
     operators::Vector  # the original list of RealFourierOperator
                        # (as many as there are terms), kept for easier exploration
     optimized_operators::Vector  # the optimized list of RealFourierOperator, to be used for applying
-    HamiltonianBlock(basis, kpt, operators) = new(basis, kpt, operators, optimize_operators_(operators))
+    preallocated  # Pre-allocated scratch arrays for fast application
```
Comment (antoine-levitt) [inline src/terms/Hamiltonian.jl:14]: call it `scratch` instead. Possibly `_scratch` to emphasize it's very internal?

**Pull request merged**